core/src/Language/Avaleryar/Semantics.hs

```diff
@@ -222,16 +223,16 @@ unifyTerm t t' = do
   unless (ts == ts') $ do
     rt@RT {..} <- getRT
     case (ts, ts') of
-      (Var v, _) -> putRT rt {env = Map.insert v ts' env}
-      (_, Var v) -> putRT rt {env = Map.insert v ts  env}
+      (Var v, _) -> putRT rt {env = Partition.joinElems (Var v) ts' env}
+      (_, Var v) -> putRT rt {env = Partition.joinElems (Var v) ts  env}
       _          -> empty -- ts /= ts', both are values
 
 -- | Apply the current substitution on the given 'Term'.  This function does path compression: if it
```

Comment:
> This function does path compression: if it finds a variable, it recurs.

This statement is half true, it does recur, but it does not update the environment to compress the path. This PR tries to address that using the union find algorithm.